crates/runtime/src/instance.rs

```diff
@@ -50,17 +50,16 @@ pub(crate) struct Instance {
     /// WebAssembly table data.
     tables: BoxedSlice<DefinedTableIndex, Table>,
 
-    /// Passive elements in this instantiation. As `elem.drop`s happen, these
-    /// entries get removed. A missing entry is considered equivalent to an
-    /// empty slice.
-    passive_elements: RefCell<HashMap<ElemIndex, Box<[*mut VMCallerCheckedAnyfunc]>>>,
+    /// Stores the dropped passive element segments in this instantiation by index.
+    /// If the index is present in the set, the segment has been dropped.
+    dropped_elements: RefCell<EntitySet<ElemIndex>>,
 
-    /// Passive data segments from our module. As `data.drop`s happen, entries
-    /// get removed. A missing entry is considered equivalent to an empty slice.
-    passive_data: RefCell<HashMap<DataIndex, Arc<[u8]>>>,
+    /// Stores the dropped passive data segments in this instantiation by index.
+    /// If the index is present in the set, the segment has been dropped.
+    dropped_data: RefCell<EntitySet<DataIndex>>,
 
     /// Hosts can store arbitrary per-instance information here.
-    host_state: Box<dyn Any>,
+    host_state: Option<Box<dyn Any>>,
```

Comment:
FWIW `Box::new(SomeZst) as Box<dyn Any>` doesn't actually allocate anything, so if the purpose of this was to eschew allocations I think it's fine to leave this without the `Option`?

Reply:
It wasn't to elide an allocation, but to take the host state out of the instance allocation request when creating the instance handle and passing the remainder of the request to `initialize_vmcontext`.

I'm happy to undo this change and just splat what's needed from the request as discrete arguments to `initialize_vmcontext`.

Reply:
Ah ok that seems fine to leave as-is in that case!

Reply:
Given the latest changes to module linking which I've now rebased on, this is back to being `Box<dyn Any>`.